File: mediapicker/src/main/java/vn/tungdx/mediapicker/MediaAdapter.java
package vn.tungdx.mediapicker;

import android.content.Context;
import android.database.Cursor;
import android.net.Uri;
import android.support.v4.widget.CursorAdapter;
import android.view.View;
import android.view.ViewGroup;
import android.view.ViewGroup.LayoutParams;
import android.widget.AbsListView.RecyclerListener;
import android.widget.RelativeLayout;

import java.util.ArrayList;
import java.util.List;

import vn.tungdx.mediapicker.imageloader.MediaImageLoader;
import vn.tungdx.mediapicker.utils.MediaUtils;
import vn.tungdx.mediapicker.widget.PickerImageView;

/**
 * @author TUNGDX
 */

/**
 * Adapter for display media item list.
 */
public class MediaAdapter extends CursorAdapter implements RecyclerListener {
    private int mMediaType;
    private MediaImageLoader mMediaImageLoader;
    private List<MediaItem> mMediaListSelected = new ArrayList<MediaItem>();
    private MediaOptions mMediaOptions;
    private int mItemHeight = 0;
    private int mNumColumns = 0;
    private RelativeLayout.LayoutParams mImageViewLayoutParams;
    private List<PickerImageView> mPickerImageViewSelected = new ArrayList<PickerImageView>();

    public MediaAdapter(Context context, Cursor c, int flags,
                        MediaImageLoader mediaImageLoader, int mediaType, MediaOptions mediaOptions) {
        this(context, c, flags, null, mediaImageLoader, mediaType, mediaOptions);
    }

    public MediaAdapter(Context context, Cursor c, int flags,
                        List<MediaItem> mediaListSelected, MediaImageLoader mediaImageLoader,
                        int mediaType, MediaOptions mediaOptions) {
        super(context, c, flags);
        if (mediaListSelected != null)
            mMediaListSelected = mediaListSelected;
        mMediaImageLoader = mediaImageLoader;
        mMediaType = mediaType;
        mMediaOptions = mediaOptions;
        mImageViewLayoutParams = new RelativeLayout.LayoutParams(
                LayoutParams.MATCH_PARENT, LayoutParams.WRAP_CONTENT);
    }

    @Override
    public void bindView(View view, Context context, Cursor cursor) {
        final ViewHolder holder = (ViewHolder) view.getTag();
        final Uri uri;
        if (mMediaType == MediaItem.PHOTO) {
            uri = MediaUtils.getPhotoUri(cursor);
            holder.thumbnail.setVisibility(View.GONE);
        } else {
            uri = MediaUtils.getVideoUri(cursor);
            holder.thumbnail.setVisibility(View.VISIBLE);
        }
        boolean isSelected = isSelected(uri);
        holder.imageView.setSelected(isSelected);
        if (isSelected) {
            mPickerImageViewSelected.add(holder.imageView);
        }
        mMediaImageLoader.displayImage(uri, holder.imageView);
    }

    @Override
    public View newView(Context context, Cursor cursor, ViewGroup viewGroup) {
        ViewHolder holder = new ViewHolder();
        View root = View
                .inflate(mContext, R.layout.list_item_mediapicker, null);
        holder.imageView = (PickerImageView) root.findViewById(R.id.thumbnail);
        holder.thumbnail = root.findViewById(R.id.overlay);

        holder.imageView.setLayoutParams(mImageViewLayoutParams);
        // Check the height matches our calculated column width
        if (holder.imageView.getLayoutParams().height != mItemHeight) {
            holder.imageView.setLayoutParams(mImageViewLayoutParams);
        }
        root.setTag(holder);
        return root;
    }

    private class ViewHolder {
        PickerImageView imageView;
        View thumbnail;
    }

    public boolean hasSelected() {
        return mMediaListSelected.size() > 0;
    }

    /**
     * Check media uri is selected or not.
     *
     * @param uri Uri of media item (photo, video)
     * @return true if selected, false otherwise.
     */
    public boolean isSelected(Uri uri) {
        if (uri == null)
            return false;
        for (MediaItem item : mMediaListSelected) {
            if (item.getUriOrigin().equals(uri))
                return true;
        }
        return false;
    }

    /**
     * Check {@link MediaItem} is selected or not.
     *
     * @param item {@link MediaItem} to check.
     * @return true if selected, false otherwise.
     */
    public boolean isSelected(MediaItem item) {
        return mMediaListSelected.contains(item);
    }

    /**
     * Set {@link MediaItem} selected.
     *
     * @param item {@link MediaItem} to selected.
     */
    public void setMediaSelected(MediaItem item) {
        syncMediaSelectedAsOptions();
        if (!mMediaListSelected.contains(item))
            mMediaListSelected.add(item);
    }

    /**
     * If item selected then change to unselected and unselected to selected.
     *
     * @param item Item to update.
     */
    public void updateMediaSelected(MediaItem item,
                                    PickerImageView pickerImageView) {
        if (mMediaListSelected.contains(item)) {
            mMediaListSelected.remove(item);
            pickerImageView.setSelected(false);
            this.mPickerImageViewSelected.remove(pickerImageView);
        } else {
            boolean value = syncMediaSelectedAsOptions();
            if (value) {
                for (PickerImageView picker : this.mPickerImageViewSelected) {
                    picker.setSelected(false);
                }
                this.mPickerImageViewSelected.clear();
            }
            mMediaListSelected.add(item);
            pickerImageView.setSelected(true);
            this.mPickerImageViewSelected.add(pickerImageView);
        }
    }

    /**
     * @return List of {@link MediaItem} selected.
     */
    public List<MediaItem> getMediaSelectedList() {
        return mMediaListSelected;
    }

    /**
     * Set list of {@link MediaItem} selected.
     *
     * @param list
     */
    public void setMediaSelectedList(List<MediaItem> list) {
        mMediaListSelected = list;
    }

    /**
     * Whether clear or not media selected as options.
     *
     * @return true if clear, false otherwise.
     */
    private boolean syncMediaSelectedAsOptions() {
        switch (mMediaType) {
            case MediaItem.PHOTO:
                if (!mMediaOptions.canSelectMultiPhoto()) {
                    mMediaListSelected.clear();
                    return true;
                }
                break;
            case MediaItem.VIDEO:
                if (!mMediaOptions.canSelectMultiVideo()) {
                    mMediaListSelected.clear();
                    return true;
                }

                break;
            default:
                break;
        }
        return false;
    }

    /**
     * {@link MediaItem#VIDEO} or {@link MediaItem#PHOTO}
     *
     * @param mediaType
     */
    public void setMediaType(int mediaType) {
        mMediaType = mediaType;
    }

    // set numcols
    public void setNumColumns(int numColumns) {
        mNumColumns = numColumns;
    }

    public int getNumColumns() {
        return mNumColumns;
    }

    // set photo item height
    public void setItemHeight(int height) {
        if (height == mItemHeight) {
            return;
        }
        mItemHeight = height;
        mImageViewLayoutParams.height = height;
        mImageViewLayoutParams.width = height;
        notifyDataSetChanged();
    }

    @Override
    public void onMovedToScrapHeap(View view) {
        PickerImageView imageView = (PickerImageView) view
                .findViewById(R.id.thumbnail);
        mPickerImageViewSelected.remove(imageView);
    }

    public void onDestroyView() {
        mPickerImageViewSelected.clear();
    }
}

File: mediapicker/src/main/java/vn/tungdx/mediapicker/MediaAdapter.kt
package vn.tungdx.mediapicker

import android.content.Context
import android.database.Cursor
import android.net.Uri
import android.view.View
import android.view.ViewGroup
import android.view.ViewGroup.LayoutParams
import android.widget.AbsListView.RecyclerListener
import android.widget.RelativeLayout
import androidx.cursoradapter.widget.CursorAdapter
import vn.tungdx.mediapicker.imageloader.MediaImageLoader
import vn.tungdx.mediapicker.utils.MediaUtils
import vn.tungdx.mediapicker.widget.PickerImageView
import java.util.*

/**
 * @author TUNGDX
 */

/**
 * Adapter for display media item list.
 */
class MediaAdapter(context: Context, c: Cursor, flags: Int,
                   mediaListSelected: MutableList<MediaItem>?, private val mMediaImageLoader: MediaImageLoader,
                   private var mMediaType: Int, private val mMediaOptions: MediaOptions) : CursorAdapter(context, c, flags), RecyclerListener {
    private var mMediaListSelected: MutableList<MediaItem> = ArrayList()
    private var mItemHeight = 0
    // set numcols
    var numColumns = 0
    private val mImageViewLayoutParams: RelativeLayout.LayoutParams
    private val mPickerImageViewSelected = ArrayList<PickerImageView>()

    /**
     * @return List of [MediaItem] selected.
     */
    /**
     * Set list of [MediaItem] selected.
     *
     * @param list
     */
    var mediaSelectedList: MutableList<MediaItem>
        get() = mMediaListSelected
        set(list) {
            mMediaListSelected = list
        }

    constructor(context: Context, c: Cursor, flags: Int,
                mediaImageLoader: MediaImageLoader, mediaType: Int, mediaOptions: MediaOptions) : this(context, c, flags, null, mediaImageLoader, mediaType, mediaOptions) {
    }

    init {
        if (mediaListSelected != null)
            mMediaListSelected = mediaListSelected
        mImageViewLayoutParams = RelativeLayout.LayoutParams(
                LayoutParams.MATCH_PARENT, LayoutParams.WRAP_CONTENT)
    }

    override fun bindView(view: View, context: Context, cursor: Cursor) {
        val holder = view.tag as ViewHolder
        val uri: Uri
        if (mMediaType == MediaItem.PHOTO) {
            uri = MediaUtils.getPhotoUri(cursor)
            holder.thumbnail!!.visibility = View.GONE
        } else {
            uri = MediaUtils.getVideoUri(cursor)
            holder.thumbnail!!.visibility = View.VISIBLE
        }
        val isSelected = isSelected(uri)
        holder.imageView!!.isSelected = isSelected
        if (isSelected) {
            mPickerImageViewSelected.add(holder.imageView!!)
        }
        mMediaImageLoader.displayImage(uri, holder.imageView!!)
    }

    override fun newView(context: Context, cursor: Cursor, viewGroup: ViewGroup): View {
        val holder = ViewHolder()
        val root = View
                .inflate(mContext, R.layout.list_item_mediapicker, null)
        holder.imageView = root.findViewById<View>(R.id.thumbnail) as PickerImageView
        holder.thumbnail = root.findViewById(R.id.overlay)

        holder.imageView!!.layoutParams = mImageViewLayoutParams
        // Check the height matches our calculated column width
        if (holder.imageView!!.layoutParams.height != mItemHeight) {
            holder.imageView!!.layoutParams = mImageViewLayoutParams
        }
        root.tag = holder
        return root
    }

    private inner class ViewHolder {
        internal var imageView: PickerImageView? = null
        internal var thumbnail: View? = null
    }

    fun hasSelected(): Boolean {
        return mMediaListSelected.size > 0
    }

    /**
     * Check media uri is selected or not.
     *
     * @param uri Uri of media item (photo, video)
     * @return true if selected, false otherwise.
     */
    fun isSelected(uri: Uri?): Boolean {
        if (uri == null)
            return false
        for (item in mMediaListSelected) {
            if (item.uriOrigin == uri)
                return true
        }
        return false
    }

    /**
     * Check [MediaItem] is selected or not.
     *
     * @param item [MediaItem] to check.
     * @return true if selected, false otherwise.
     */
    fun isSelected(item: MediaItem): Boolean {
        return mMediaListSelected.contains(item)
    }

    /**
     * Set [MediaItem] selected.
     *
     * @param item [MediaItem] to selected.
     */
    fun setMediaSelected(item: MediaItem) {
        syncMediaSelectedAsOptions()
        if (!mMediaListSelected.contains(item))
            mMediaListSelected.add(item)
    }

    /**
     * If item selected then change to unselected and unselected to selected.
     *
     * @param item Item to update.
     */
    fun updateMediaSelected(item: MediaItem,
                            pickerImageView: PickerImageView) {
        if (mMediaListSelected.contains(item)) {
            mMediaListSelected.remove(item)
            pickerImageView.isSelected = false
            this.mPickerImageViewSelected.remove(pickerImageView)
        } else {
            val value = syncMediaSelectedAsOptions()
            if (value) {
                for (picker in this.mPickerImageViewSelected) {
                    picker.isSelected = false
                }
                this.mPickerImageViewSelected.clear()
            }
            mMediaListSelected.add(item)
            pickerImageView.isSelected = true
            this.mPickerImageViewSelected.add(pickerImageView)
        }
    }

    /**
     * Whether clear or not media selected as options.
     *
     * @return true if clear, false otherwise.
     */
    private fun syncMediaSelectedAsOptions(): Boolean {
        when (mMediaType) {
            MediaItem.PHOTO -> if (!mMediaOptions.canSelectMultiPhoto()) {
                mMediaListSelected.clear()
                return true
            }
            MediaItem.VIDEO -> if (!mMediaOptions.canSelectMultiVideo()) {
                mMediaListSelected.clear()
                return true
            }
            else -> {
            }
        }
        return false
    }

    /**
     * [MediaItem.VIDEO] or [MediaItem.PHOTO]
     *
     * @param mediaType
     */
    fun setMediaType(mediaType: Int) {
        mMediaType = mediaType
    }

    // set photo item height
    fun setItemHeight(height: Int) {
        if (height == mItemHeight) {
            return
        }
        mItemHeight = height
        mImageViewLayoutParams.height = height
        mImageViewLayoutParams.width = height
        notifyDataSetChanged()
    }

    override fun onMovedToScrapHeap(view: View) {
        val imageView = view
                .findViewById<View>(R.id.thumbnail) as PickerImageView
        mPickerImageViewSelected.remove(imageView)
    }

    fun onDestroyView() {
        mPickerImageViewSelected.clear()
    }
}